sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py

```diff
@@ -70,20 +72,26 @@ def test_derivatives(loss, x0, y_true):
     x0 = np.array([x0], dtype=Y_DTYPE).reshape(1, 1)
     get_gradients, get_hessians = get_derivatives_helper(loss)
 
-    def func(x):
-        return loss.pointwise_loss(y_true, x)
+    def func(x: float) -> float:
+        if isinstance(loss, _LOSSES['binary_crossentropy']):
+            # Subtract a constant term such that the binary cross entropy
+            # has its minimum at zero. This only works if 0 < y_true < 1.
+            return (loss.pointwise_loss(y_true, x)
+                    - loss.pointwise_loss(y_true, logit(y_true))).item(0)
+        else:
+            return loss.pointwise_loss(y_true, x).item(0)
 
-    def fprime(x):
-        return get_gradients(y_true, x)
+    def fprime(x: float) -> float:
```

Comment:
> I guess we need a more global decision on whether to use type annotations

I think the current situation is outlined in https://github.com/scikit-learn/scikit-learn/blame/fb76de72e7560aaa739872e94bee761777e54c0a/doc/developers/contributing.rst#L426

I did mention it in a dev meeting a month or two ago and didn't hear much objections. Again the idea is not to type everything, but now and then it helps readability. I think we can keep these in, they can't hurt in any case.

Reply:
Here in particular I find that without type annotations (and docstrings), it's very difficult to know, what's the expected input and output of,
```py
    def fprime(x):
        return get_gradients(y_true, x).item(0)
```

Reply:
Ok.

In this specific case, I think we should actually let numpy deal with arrays in these functions (comment just below), so we'd need to remove the annotations anyway

Reply:
Also, the input is actually not a float, it's a 2d array (with only one element).

I really think we should just let numpy and scipy properly deal with broadcasting here, and only ravel the arrays at the very end.

Reply:
The argument `x` of `def func(x)` is the `x0` of `@pytest.mark.parametrize('loss, x0...)` and is actually a float and not a ndarray. Same for `fpime` and `fprime2`.
AFAIU, the `newton` root finder does prefer scalars, not ndarrays as those have different meaning.

I'll do whatever you prefer here. It's just not clear to me what that is 😏

Reply:
> AFAIU, the newton root finder does prefer scalars

That's true, but it does fine with arrays with a single element

We need to wrap x0 and y_true into arrays because that's what the loss methods expect. `newton` is OK with these. If we're going to recast these 1-element arrays into floats, it's cleaner and easier to do it once at the very end, i.e. by calling `.ravel` just before the assertions.

Alternatively, we can also go back to just using `assert np.allclose`. I know we sort of prefer `assert_allclose`, but at the end of the day, that change is unrelated to this PR.